seed/seedwriter/writer.go

```diff
@@ -302,49 +340,118 @@ func (w *Writer) SetOptionsSnaps(optSnaps []*OptionSnap) error {
 	if err := w.checkStep(startStep); err != nil {
-		return err
+		return nil, err
 	}
 	if db == nil {
-		return fmt.Errorf("internal error: Writer *asserts.RODatabsae is nil")
+		return nil, fmt.Errorf("internal error: Writer *asserts.RODatabase is nil")
 
 	}
 	if newFetcher == nil {
-		return fmt.Errorf("internal error: Writer newFetcherFunc is nil")
+		return nil, fmt.Errorf("internal error: Writer newFetcherFunc is nil")
 	}
 	w.db = db
 
 	f := MakeRefAssertsFetcher(newFetcher)
 
 	// XXX support UBUNTU_IMAGE_SKIP_COPY_UNVERIFIED_MODEL ?
 	if err := f.Save(w.model); err != nil {
-		return fmt.Errorf("cannot fetch and check prerequisites for the model assertion: %v", err)
+		return nil, fmt.Errorf("cannot fetch and check prerequisites for the model assertion: %v", err)
 	}
 
 	w.modelRefs = f.Refs()
 
 	// XXX get if needed the store assertion
 
-	return w.tree.mkFixedDirs()
+	if err := w.tree.mkFixedDirs(); err != nil {
+		return nil, err
+	}
+
+	return f, nil
 }
 
-// LocalSnaps()
-// InfoDerived()
+// LocalSnaps returns a list of seed snaps that are local.  The writer
+// delegates to produce *snap.Info for them to then be set via
+// SetInfo. If matching snap assertions can be found as well they can
+// be passed into SeedSnap ARefs, assuming they were added to the
+// writing assertion database.
+func (w *Writer) LocalSnaps() ([]*SeedSnap, error) {
+	if err := w.checkStep(localSnapsStep); err != nil {
+		return nil, err
+	}
+
+	return w.localSnaps, nil
+}
+
+// InfoDerived checks the local snaps metadata provided via setting it
+// into the SeedSnaps returned by the previous LocalSnaps.
+func (w *Writer) InfoDerived() error {
+	if err := w.checkStep(infoDerivedStep); err != nil {
+		return err
+	}
+
+	for _, sn := range w.localSnaps {
+		if sn.Info == nil {
+			return fmt.Errorf("internal error: at this point snap %q Info should have been set", sn.Path)
+		}
+		sn.SnapRef = sn.Info
+
+		// local snap gets local revision
+		if sn.Info.Revision.Unset() {
+			sn.Info.Revision = snap.R(-1)
+		}
+
+		if w.byRefLocalSnaps.Contains(sn) {
+			return fmt.Errorf("local snap %q is repeated in options", sn.SnapName())
+		}
+
+		// in case, merge channel given by name separately
+		optSnap, _ := w.byNameOptSnaps.Lookup(sn).(*OptionSnap)
+		if optSnap != nil && optSnap.Channel != "" {
+			if sn.optionSnap.Channel != "" {
+				if sn.optionSnap.Channel != optSnap.Channel {
+					return fmt.Errorf("option snap has different channels specified: %q=%q vs %q=%q", sn.Path, sn.optionSnap.Channel, optSnap.Name, optSnap.Channel)
+				}
+			} else {
+				sn.optionSnap.Channel = optSnap.Channel
+			}
+		}
+
+		w.byRefLocalSnaps.Add(sn)
+	}
+
+	return nil
+}
 
 // SetInfo sets Info of the SeedSnap and possibly computes its
 // destination Path.
 func (w *Writer) SetInfo(sn *SeedSnap, info *snap.Info) error {
+	sn.Info = info
 	if sn.local {
-		panic("NOT IMPLEMENTED YET")
+		// nothing more to do
```

Comment:
I'm probably missing something, but don't we need to set sn.Path below for local snap?

Reply:
For local snaps the Path is set to their origin Path when we build the SeedSnap:

https://github.com/snapcore/snapd/pull/7468/files#diff-86bf37ed48cf745a5d445a22170384c2R346

We keep it like that until we copy them into the seed and then remember the final in-seed path:

https://github.com/snapcore/snapd/pull/7468/files#diff-86bf37ed48cf745a5d445a22170384c2R683

For store snaps we set the Path below so to the tell the caller where it should download them on behalf of the Writer.

I considered having different path fields for a bit more clarity, it seemed a bit overkill though.

Reply:
Understood, thanks.